File: Dante2/Module2.02/Zadanie2.06/main.c
#include <stdio.h>
#include <math.h>

struct point_t
{
    int x;
    int y;
};

struct segment_t
{
    struct point_t start;
    struct point_t end;
};

struct segment_t* read_segment(struct segment_t* p);
void show_segment(const struct segment_t* p);
float length(const struct segment_t* s, int *err_code);

int main()
{
    struct segment_t p;
    printf("Enter: ");

    if(read_segment(&p) == NULL)
    {
        printf("Incorrect input");
        return 1;
    }

    show_segment(&p);

    int err_code;
    float result = length(&p, &err_code);
    if(result == -1.0f)
    {
        printf("Incorrect input");
        return 1;
    }

    printf("%.2f",result);

    return 0;
}

float length(const struct segment_t* s, int *err_code)
{
    if (s == NULL)
    {
        if (err_code == NULL)
        {
            return -1.0f;
        }
        *err_code = 1;
        return -1.0f;
    }

    float distance = sqrtf((powf((float)s->end.x - (float)s->start.x, 2)) + (powf((float)s->end.y - (float)s->start.y, 2)));
    if (err_code == NULL)
    {
        return distance;
    }

    *err_code = 0;
    return distance;
}

void show_segment(const struct segment_t* p)
{
    if(p == NULL)
    {
        return;
    }
    printf("%i; %i\n",p->start.x, p->start.y);
    printf("%i; %i\n",p->end.x, p->end.y);
}

struct segment_t* read_segment(struct segment_t* p)
{
    if(p == NULL)
    {
        return NULL;
    }

    int error;

    printf("Enter x1: ");
    error = scanf("%i",&p->start.x);
    if(!error)
    {
        return NULL;
    }
    printf("Enter y1: ");
    error = scanf("%i",&p->start.y);
    if(!error)
    {
        return NULL;
    }
    printf("Enter x2: ");
    error = scanf("%i",&p->end.x);
    if(!error)
    {
        return NULL;
    }
    printf("Enter y2: ");
    error = scanf("%i",&p->end.y);
    if(!error)
    {
        return NULL;
    }

    return p;
}

File: Dante2/Module2.02/Zadanie2.06/main.rs
#[derive(Default)]
struct PointT {
    x: i32,
    y: i32,
}
#[derive(Default)]
struct SegmentT {
    start: PointT,
    end: PointT,
}
fn main() {
    let mut p: SegmentT = Default::default();
    read_segment(&mut p);
    show_segment(&p);

    let result: f32 = length(&p);
    print!("{}", result);
}

fn read_segment(p: &mut SegmentT) -> & SegmentT {
    let mut input: String = String::new();

    println!("Enter: x1");
    input.clear();
    std::io::stdin().read_line(&mut input).expect("Incorrect");
    p.start.x = input.trim().parse().expect("Incorrect");

    println!("Enter: y1");
    input.clear();
    std::io::stdin().read_line(&mut input).expect("Incorrect");
    p.start.y = input.trim().parse().expect("Incorrect");

    println!("Enter: x2");
    input.clear();
    std::io::stdin().read_line(&mut input).expect("Incorrect");
    p.end.x = input.trim().parse().expect("Incorrect");

    println!("Enter: y2");
    input.clear();
    std::io::stdin().read_line(&mut input).expect("Incorrect");
    p.end.y = input.trim().parse().expect("Incorrect");

    p
}

fn show_segment(p: & SegmentT) {
    println!("{}; {}", p.start.x, p.start.y);
    println!("{}; {}", p.end.x, p.end.y);
}

fn length(p: & SegmentT) -> f32 {
    let distance: f32 = f32::sqrt((f32::powi((p.end.x - p.start.x) as f32, 2)) +
        (f32::powi((p.end.y - p.start.y) as f32, 2)));
    distance
}
